libs/httpproxy/httpproxy.go

```diff
@@ -0,0 +1,142 @@
+package httpproxy
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+type Proxy struct {
+	ctx            context.Context
+	targetURL      *url.URL
+	client         *http.Client
+	server         *http.Server
+	headerToInject map[string]string
+}
+
+// NewProxy creates a new proxy instance that will forward all requests to the targetURL
+// The targetURL should be a valid URL with a scheme and a host.
+func NewProxy(targetURL string) (*Proxy, error) {
+	u, err := url.Parse(targetURL)
+	if err != nil {
+		return nil, err
+	}
+	proxy := Proxy{targetURL: u, client: &http.Client{}, ctx: context.Background()}
+	server := &http.Server{}
+	server.Handler = http.HandlerFunc(proxy.proxyHandler)
+	proxy.server = server
+	return &proxy, nil
+}
+
+// Start starts the proxy server on the given address (host:port, e.g. localhost:8080)
+// The proxy will forward all requests to the targetURL
+func (p *Proxy) Start(addr string) error {
+	p.server.Addr = addr
+	return p.server.ListenAndServe()
+}
+
+func (p *Proxy) Stop() error {
+	return p.server.Shutdown(p.ctx)
+}
+
+// InjectHeader injects a header that will be added to all requests forwarded by the proxy
+func (p *Proxy) InjectHeader(key, value string) {
+	if p.headerToInject == nil {
+		p.headerToInject = make(map[string]string)
+	}
+	p.headerToInject[key] = value
+}
+
+func (p *Proxy) proxyHandler(w http.ResponseWriter, r *http.Request) {
+	if p.isWebSocketRequest(r) {
+		p.handleWebSocket(w, r)
```

Comment:
I'm not sure this complies with the HTTP spec. The spec mentions that the HTTP response should be a 101 status code and then the communication can happen over websocket.

ref: https://arc.net/l/quote/jacwykww

Reply:
Not sure I understand, what exactly not complies? handleWebSocket just introduces the hijacker and proxies requests back and forth, it doesn't do anything else that changes the way connection is established